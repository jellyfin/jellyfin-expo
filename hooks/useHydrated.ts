/**
 * Copyright (c) 2025 Jellyfin Contributors
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { useEffect, useState } from 'react';

import { useDownloadStore } from '../stores/DownloadStore';
import { useRootStore } from '../stores/RootStore';
import { useServerStore } from '../stores/ServerStore';
import { useSettingStore } from '../stores/SettingStore';

export const useIsHydrated = () => {
	const [ isHydrated, setIsHydrated ] = useState(false);
	const [ isRootStoreHydrated, setIsRootStoreHydrated ] = useState(false);
	const [ isDownloadStoreHydrated, setIsDownloadStoreHydrated ] = useState(false);
	const [ isServerStoreHydrated, setIsServerStoreHydrated ] = useState(false);
	const [ isSettingStoreHydrated, setIsSettingStoreHydrated ] = useState(false);

	useEffect(() => {
		// Root Store
		const unsubRootFinishHydration = useRootStore.persist.onFinishHydration(() => setIsRootStoreHydrated(true));
		setIsRootStoreHydrated(useRootStore.persist.hasHydrated());
		// Download Store
		const unsubDownloadFinishHydration = useDownloadStore.persist.onFinishHydration(() => setIsDownloadStoreHydrated(true));
		setIsDownloadStoreHydrated(useDownloadStore.persist.hasHydrated());
		// Server Store
		const unsubServerFinishHydration = useServerStore.persist.onFinishHydration((state) => {
			console.debug('>>> hydrated state', state);
			setIsServerStoreHydrated(true);
		});
		setIsServerStoreHydrated(useServerStore.persist.hasHydrated());
		// Settings Store
		const unsubSettingFinishHydration = useSettingStore.persist.onFinishHydration(() => setIsSettingStoreHydrated(true));
		setIsSettingStoreHydrated(useSettingStore.persist.hasHydrated());

		return () => {
			unsubRootFinishHydration();
			unsubDownloadFinishHydration();
			unsubServerFinishHydration();
			unsubSettingFinishHydration();
		};
	}, []);

	useEffect(() => {
		console.debug('Hydration state', isRootStoreHydrated, isDownloadStoreHydrated, isServerStoreHydrated, isSettingStoreHydrated);
		if (isRootStoreHydrated && isDownloadStoreHydrated && isServerStoreHydrated && isSettingStoreHydrated) {
			setIsHydrated(true);
		}
	}, [ isDownloadStoreHydrated, isRootStoreHydrated, isServerStoreHydrated, isSettingStoreHydrated ]);

	return isHydrated;
};
