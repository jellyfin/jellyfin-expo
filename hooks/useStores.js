/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
import { useDownloadStore } from '../stores/DownloadStore';
import { useMediaStore } from '../stores/MediaStore';
import { useRootStore } from '../stores/RootStore';
import { useServerStore } from '../stores/ServerStore';
import { useSettingStore } from '../stores/SettingStore';

// Compatibility for zustand conversion
export const useStores = () => ({
	rootStore: useRootStore(),
	downloadStore: useDownloadStore(),
	serverStore: useServerStore(),
	mediaStore: useMediaStore(),
	settingStore: useSettingStore()
});
