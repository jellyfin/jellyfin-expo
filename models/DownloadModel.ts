/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

import * as FileSystem from 'expo-file-system';
import { v4 as uuidv4 } from 'uuid';

export default class DownloadModel {
	isComplete = false
	isDownloading = false
	isNew = true

	apiKey: string
	itemId: string
	/** The "play" session ID for reporting a download has stopped. */
	sessionId = uuidv4()
	serverId: string
	serverUrl: string

	title: string
	filename: string

	downloadUrl: string

	constructor(
		itemId: string,
		serverId: string,
		serverUrl: string,
		apiKey: string,
		title: string,
		filename: string,
		downloadUrl: string
	) {
		this.itemId = itemId;
		this.serverId = serverId;
		this.serverUrl = serverUrl;
		this.apiKey = apiKey;
		this.title = title;
		this.filename = filename;
		this.downloadUrl = downloadUrl;
	}

	get key() {
		return `${this.serverId}_${this.itemId}`;
	}

	get localFilename() {
		return this.filename.slice(0, this.filename.lastIndexOf('.')) + '.mp4';
	}

	get localPath() {
		return `${FileSystem.documentDirectory}${this.serverId}/${this.itemId}/`;
	}

	get uri() {
		return this.localPath + encodeURI(this.localFilename);
	}

	getStreamUrl(deviceId: string, params?: Record<string, string>): URL {
		const streamParams = new URLSearchParams({
			deviceId,
			api_key: this.apiKey,
			playSessionId: this.sessionId,
			// TODO: add mediaSourceId to support alternate media versions
			videoCodec: 'hevc,h264',
			audioCodec: 'aac,mp3,ac3,eac3,flac,alac',
			maxAudioChannels: '6',
			// subtitleCodec: 'srt,vtt',
			// subtitleMethod: 'Encode',
			...params
		});
		return new URL(`${this.serverUrl}Videos/${this.itemId}/stream.mp4?${streamParams.toString()}`);
	}
}
