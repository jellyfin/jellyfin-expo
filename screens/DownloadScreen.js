/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

import { useFocusEffect, useNavigation } from '@react-navigation/native';
import * as FileSystem from 'expo-file-system';
import React, { useCallback, useContext, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { Alert, FlatList, StyleSheet } from 'react-native';
import { Button, ThemeContext } from 'react-native-elements';
import { SafeAreaView } from 'react-native-safe-area-context';

import DownloadListItem from '../components/DownloadListItem';
import MediaTypes from '../constants/MediaTypes';
import { useStores } from '../hooks/useStores';

const DownloadScreen = () => {
	const navigation = useNavigation();
	const { downloadStore, mediaStore } = useStores();
	const { t } = useTranslation();
	const { theme } = useContext(ThemeContext);
	const [ isEditMode, setIsEditMode ] = useState(false);
	const [ selectedItems, setSelectedItems ] = useState([]);

	function exitEditMode() {
		setIsEditMode(false);
		setSelectedItems([]);
	}

	React.useLayoutEffect(() => {
		async function deleteItem(download) {
			// TODO: Add user messaging on errors
			try {
				await FileSystem.deleteAsync(download.localPath);
				downloadStore.downloads.delete(download.key);
				console.log('[DownloadScreen] download "%s" deleted', download.title);
			} catch (e) {
				console.error('[DownloadScreen] Failed to delete download', e);
			}
		}

		function onDeleteItems(downloads) {
			Alert.alert(
				'Delete Downloads',
				'These items will be permanently deleted from this device.',
				[
					{
						text: t('common.cancel'),
						onPress: exitEditMode
					},
					{
						text: `Delete ${downloads.length} Downloads`,
						onPress: async () => {
							await Promise.all(downloads.map(deleteItem));
							exitEditMode();
						},
						style: 'destructive'
					}
				]
			);
		}

		navigation.setOptions({
			headerLeft: () => (
				isEditMode ?
					<Button
						title={t('common.cancel')}
						type='clear'
						onPress={exitEditMode}
						style={styles.leftButton}
					/> :
					null
			),
			headerRight: () => (
				isEditMode ?
					<Button
						title={t('common.delete')}
						type='clear'
						style={styles.rightButton}
						disabled={selectedItems.length < 1}
						onPress={() => {
							onDeleteItems(selectedItems);
						}}
					/> :
					<Button
						title={t('common.edit')}
						type='clear'
						style={styles.rightButton}
						disabled={downloadStore.downloads.size < 1}
						onPress={() => {
							setIsEditMode(true);
						}}
					/>
			)
		});
	}, [ navigation, isEditMode, selectedItems, downloadStore.downloads ]);

	useFocusEffect(
		useCallback(() => {
			downloadStore.downloads
				.forEach(download => {
					if (download.isNew) {
						download.isNew = !download.isComplete;
					}
				});
		}, [ downloadStore.downloads ])
	);

	const downloadList = [];
	console.log('downloads', downloadStore.downloads);
	downloadStore.downloads.forEach(download => downloadList.push(download));

	return (
		<SafeAreaView
			style={{
				...styles.container,
				backgroundColor: theme.colors.background
			}}
			edges={[ 'right', 'left' ]}
		>
			<FlatList
				data={downloadList}
				extraData={downloadStore.downloads}
				renderItem={({ item, index }) => (
					<DownloadListItem
						item={item}
						index={index}
						isEditMode={isEditMode}
						isSelected={selectedItems.includes(item)}
						onSelect={() => {
							if (selectedItems.includes(item)) {
								setSelectedItems(selectedItems.filter(selected => selected !== item));
							} else {
								setSelectedItems([ ...selectedItems, item ]);
							}
						}}
						onPlay={async () => {
							item.isNew = false;
							mediaStore.set({
								isLocalFile: true,
								type: MediaTypes.Video,
								uri: item.uri
							});
						}}
					/>
				)}
				keyExtractor={(item, index) => `download-${index}-${item.key}`}
				contentContainerStyle={styles.listContainer}
			/>
		</SafeAreaView>
	);
};

const styles = StyleSheet.create({
	container: {
		flex: 1
	},
	listContainer: {
		marginTop: 1
	},
	leftButton: {
		marginLeft: 8
	},
	rightButton: {
		marginRight: 8
	}
});

export default DownloadScreen;
