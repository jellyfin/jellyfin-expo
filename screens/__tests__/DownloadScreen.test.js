/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * @jest-environment jsdom
 * @jest-environment-options {"url": "https://jestjs.io/"}
 */

import { NavigationContainer } from '@react-navigation/native';
import { act, render } from '@testing-library/react-native';
import React from 'react';

import { useStores } from '../../hooks/useStores';
import DownloadModel from '../../models/DownloadModel';
import DownloadScreen from '../DownloadScreen';

const mockSetOptions = jest.fn();
jest.mock('@react-navigation/native', () => {
	const actualNav = jest.requireActual('@react-navigation/native');
	return {
		...actualNav,
		useNavigation: () => ({
			setOptions: mockSetOptions
		})
	};
});

const DOWNLOAD_1 = new DownloadModel(
	'item-id',
	'server-id',
	'https://example.com/',
	'api-key',
	'title',
	'file name.mkv',
	'https://example.com/download'
);

const DOWNLOAD_2 = new DownloadModel(
	'item-id-2',
	'server-id',
	'https://test2.example.com/',
	'api-key',
	'other title',
	'other file name.mkv',
	'https://test2.example.com/download'
);

const mockDownloadStore = {
	downloads: new Map([[ DOWNLOAD_1.key, DOWNLOAD_1 ], [ DOWNLOAD_2.key, DOWNLOAD_2 ]]),
	add: (v) => act(() => { mockDownloadStore.downloads = new Map([ ...mockDownloadStore.downloads, [ v.key, v ]]); })
};

jest.mock('../../hooks/useStores');
useStores.mockImplementation(() => ({
	rootStore: {},
	downloadStore: mockDownloadStore
}));

describe('DownloadScreen', () => {
	it('should render correctly', () => {
		const v = render(
			<NavigationContainer>
				<DownloadScreen />
			</NavigationContainer>
		);

		expect(v.toJSON()).toMatchSnapshot();
		expect(mockSetOptions).toHaveBeenCalled();
	});
});
