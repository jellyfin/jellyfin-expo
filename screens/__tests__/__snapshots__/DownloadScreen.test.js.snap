// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadScreen should render correctly 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "off",
      "left": "additive",
      "right": "additive",
      "top": "off",
    }
  }
  style={
    {
      "backgroundColor": undefined,
      "flex": 1,
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      {
        "marginTop": 1,
      }
    }
    data={
      [
        DownloadModel {
          "apiKey": "api-key",
          "downloadUrl": "https://example.com/download",
          "filename": "file name.mkv",
          "isComplete": false,
          "isDownloading": false,
          "isNew": true,
          "itemId": "item-id",
          "serverId": "server-id",
          "serverUrl": "https://example.com/",
          "sessionId": "uuid-0",
          "title": "title",
        },
        DownloadModel {
          "apiKey": "api-key",
          "downloadUrl": "https://test2.example.com/download",
          "filename": "other file name.mkv",
          "isComplete": false,
          "isDownloading": false,
          "isNew": true,
          "itemId": "item-id-2",
          "serverId": "server-id",
          "serverUrl": "https://test2.example.com/",
          "sessionId": "uuid-1",
          "title": "other title",
        },
      ]
    }
    extraData={
      Map {
        "server-id_item-id" => DownloadModel {
          "apiKey": "api-key",
          "downloadUrl": "https://example.com/download",
          "filename": "file name.mkv",
          "isComplete": false,
          "isDownloading": false,
          "isNew": true,
          "itemId": "item-id",
          "serverId": "server-id",
          "serverUrl": "https://example.com/",
          "sessionId": "uuid-0",
          "title": "title",
        },
        "server-id_item-id-2" => DownloadModel {
          "apiKey": "api-key",
          "downloadUrl": "https://test2.example.com/download",
          "filename": "other file name.mkv",
          "isComplete": false,
          "isDownloading": false,
          "isNew": true,
          "itemId": "item-id-2",
          "serverId": "server-id",
          "serverUrl": "https://test2.example.com/",
          "sessionId": "uuid-1",
          "title": "other title",
        },
      }
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View>
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#ffffff",
                "borderBottomWidth": 0.5,
                "borderColor": "#bcbbc1",
                "borderTopWidth": 0.5,
                "flexDirection": "row",
                "padding": 14,
              }
            }
            testID="padView"
          >
            <View
              style={
                [
                  {
                    "alignItems": "flex-start",
                    "flex": 1,
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
              theme={
                {
                  "colors": {
                    "black": "#242424",
                    "disabled": "hsl(208, 8%, 90%)",
                    "divider": "#bcbbc1",
                    "error": "#ff190c",
                    "grey0": "#393e42",
                    "grey1": "#43484d",
                    "grey2": "#5e6977",
                    "grey3": "#86939e",
                    "grey4": "#bdc6cf",
                    "grey5": "#e1e8ee",
                    "greyOutline": "#bbb",
                    "platform": {
                      "android": {
                        "error": "#f44336",
                        "grey": "rgba(0, 0, 0, 0.54)",
                        "primary": "#2196f3",
                        "searchBg": "#dcdce1",
                        "secondary": "#9C27B0",
                        "success": "#4caf50",
                        "warning": "#ffeb3b",
                      },
                      "default": {
                        "error": "#ff3b30",
                        "grey": "#7d7d7d",
                        "primary": "#007aff",
                        "searchBg": "#dcdce1",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                      "ios": {
                        "error": "#ff3b30",
                        "grey": "#7d7d7d",
                        "primary": "#007aff",
                        "searchBg": "#dcdce1",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                      "web": {
                        "error": "#ff190c",
                        "grey": "#393e42",
                        "primary": "#2089dc",
                        "searchBg": "#303337",
                        "secondary": "#ca71eb",
                        "success": "#52c41a",
                        "warning": "#faad14",
                      },
                    },
                    "primary": "#2089dc",
                    "searchBg": "#303337",
                    "secondary": "#ca71eb",
                    "success": "#52c41a",
                    "warning": "#faad14",
                    "white": "#ffffff",
                  },
                }
              }
            >
              <Text
                style={
                  {
                    "backgroundColor": "transparent",
                    "color": "#242424",
                    "fontSize": 17,
                  }
                }
                testID="title"
              >
                title
              </Text>
              <Text
                style={
                  {
                    "backgroundColor": "transparent",
                    "color": "#242424",
                    "fontSize": 15,
                  }
                }
                testID="subtitle"
              >
                file name.mp4
              </Text>
            </View>
            <View
              style={
                {
                  "paddingLeft": 16,
                }
              }
            />
            <ActivityIndicator
              testID="loading-indicator"
            />
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View>
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#ffffff",
                "borderBottomWidth": 0.5,
                "borderColor": "#bcbbc1",
                "flexDirection": "row",
                "padding": 14,
              }
            }
            testID="padView"
          >
            <View
              style={
                [
                  {
                    "alignItems": "flex-start",
                    "flex": 1,
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
              theme={
                {
                  "colors": {
                    "black": "#242424",
                    "disabled": "hsl(208, 8%, 90%)",
                    "divider": "#bcbbc1",
                    "error": "#ff190c",
                    "grey0": "#393e42",
                    "grey1": "#43484d",
                    "grey2": "#5e6977",
                    "grey3": "#86939e",
                    "grey4": "#bdc6cf",
                    "grey5": "#e1e8ee",
                    "greyOutline": "#bbb",
                    "platform": {
                      "android": {
                        "error": "#f44336",
                        "grey": "rgba(0, 0, 0, 0.54)",
                        "primary": "#2196f3",
                        "searchBg": "#dcdce1",
                        "secondary": "#9C27B0",
                        "success": "#4caf50",
                        "warning": "#ffeb3b",
                      },
                      "default": {
                        "error": "#ff3b30",
                        "grey": "#7d7d7d",
                        "primary": "#007aff",
                        "searchBg": "#dcdce1",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                      "ios": {
                        "error": "#ff3b30",
                        "grey": "#7d7d7d",
                        "primary": "#007aff",
                        "searchBg": "#dcdce1",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                      "web": {
                        "error": "#ff190c",
                        "grey": "#393e42",
                        "primary": "#2089dc",
                        "searchBg": "#303337",
                        "secondary": "#ca71eb",
                        "success": "#52c41a",
                        "warning": "#faad14",
                      },
                    },
                    "primary": "#2089dc",
                    "searchBg": "#303337",
                    "secondary": "#ca71eb",
                    "success": "#52c41a",
                    "warning": "#faad14",
                    "white": "#ffffff",
                  },
                }
              }
            >
              <Text
                style={
                  {
                    "backgroundColor": "transparent",
                    "color": "#242424",
                    "fontSize": 17,
                  }
                }
                testID="title"
              >
                other title
              </Text>
              <Text
                style={
                  {
                    "backgroundColor": "transparent",
                    "color": "#242424",
                    "fontSize": 15,
                  }
                }
                testID="subtitle"
              >
                other file name.mp4
              </Text>
            </View>
            <View
              style={
                {
                  "paddingLeft": 16,
                }
              }
            />
            <ActivityIndicator
              testID="loading-indicator"
            />
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RNCSafeAreaView>
`;
